✨ UI Rewrite for new version of Luminare

MrKai77 opened the pull request:
This branch of Loop aims to rewrite the UI to fully support the new version of Luminare, which will be merged in the near future. This includes performance improvements, logical improvements and new UI features :)

On top of that, I have also taken advantage of this opportunity to improve the overall structure of Loop's settings UI architecture, which now directly uses `Defaults`'s `@Default` property wrapper when available. This will ensure that the new UI is more reactive to certain settings the user may change. Furthermore, icons have been renamed, and the Xcode project now uses folders instead of groups.

cc @KrLite, who inspired and helped the process of rewriting of Luminare :)

MrKai77 committed:
✨ Convert Xcode groups to folders

MrKai77 committed:
🚛 Rename icons

MrKai77 committed:
✨ About view

MrKai77 committed:
✨Working Luminare v2 adaptation!

MrKai77 committed:
✨ More fixes

MrKai77 committed:
🐞 Expand hit area for locked icons to full content box

MrKai77 committed:
💄 Use `LuminareButtonCompose` for permissions buttons

MrKai77 committed:
✨ More small improvements

MrKai77 committed:
🐞 Bug fixes

MrKai77 committed:
✨ Latest Luminare update

MrKai77 committed:
Merge branch 'develop' into luminare-v2

# Conflicts:
#	Loop.xcodeproj/project.pbxproj
#	Loop/Luminare/Loop/AdvancedConfiguration.swift
#	Loop/Utilities/WallpaperProcessor.swift

MrKai77 committed:
✨ Fix assets

MrKai77 committed:
✨ Latest changes

MrKai77 committed:
🔀 Merge branch 'develop' into luminare-v2

# Conflicts:
#	Loop.xcodeproj/project.pbxproj
#	Loop/Luminare/Settings/Keybindings/KeybindsConfigurationView.swift

MrKai77 committed:
✨ Update branch for latest changes

MrKai77 committed:
🔀 Merge branch 'develop' into 'luminare-v2'

# Conflicts:
#	Loop.xcodeproj/project.pbxproj

MrKai77 committed:
✨ Small alignment fixes + Update Luminare

MrKai77 committed:
🐞 Fix padding configuration modal

MrKai77 committed:
💄 Improve textbox appearance

MrKai77 committed:
🔀 Merge branch 'develop' into luminare-v2

# Conflicts:
#	Loop/Luminare/Settings/Behavior/BehaviorConfiguration.swift
#	Loop/Luminare/Settings/Keybindings/KeybindItemView.swift

MrKai77 committed:
✨ Update to latest code & Luminare

MrKai77 committed:
✨ Update `StashActionConfigurationView`

MrKai77 committed:
✨ Copied indicator in About page

MrKai77 committed:
💄 Improved popover spacing

MrKai77 committed:
⬆️ Update Luminare

MrKai77 committed:
Fix TODOs

MrKai77 committed:
✨ Command + delete to delete keybinds and excluded apps

MrKai77 committed:
🐞 Fix small alignment issue

SenpaiHunters commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "Luminare"]
```
Is this needed, a `.gitmodules` file that is?

SenpaiHunters commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -7,7 +7,7 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
-		A860324F2CB34A6C005742EB /* Luminare in Frameworks */ = {isa = PBXBuildFile; productRef = A860324E2CB34A6C005742EB /* Luminare */; };
```
Luminares branch needs to be changed from local to it's remote, currently it's trying to find it from your local path

SenpaiHunters commented on Loop/Luminare/Settings/Keybindings/KeybindsConfigurationView.swift:
```diff
@@ -34,51 +18,77 @@ struct KeybindsConfigurationView: View {
-                CycleBackwardsView(triggerKey: triggerKey, isOn: $model.cycleBackwardsOnShiftPressed)
+                LuminareToggle(isOn: $cycleBackwardsOnShiftPressed) {
+                    Text("Cycle backward with Shift")
+                        .luminarePopover(hidden: !isShiftUsedByTriggerKey) {
```
You cannot see this preview or text when loading loop, refer to image
<img width="758" height="638" alt="006396-07@2x" src="https://github.com/user-attachments/assets/58e11148-694a-465e-9ca6-982a76f25f11" />

SenpaiHunters replied:
this is half fixed, refer to my `Hi, Kai!` comment for more info on why this happens.

SenpaiHunters commented on Loop/Luminare/Settings/Keybindings/KeybindsConfigurationView.swift:
```diff
@@ -34,51 +18,77 @@ struct KeybindsConfigurationView: View {
-                keybinds.insert(.init(.noAction), at: 0)
-            },
-            content: { keybind in
+        LuminareSection("Keybinds") {
```
When pressing esc on this stack, the popup will flash white before removing.

mrkai77 replied:
Can't reproduce this on my side. Mind sending a recording as well as your macOS version?

SenpaiHunters replied:
It's a minor thing and only displays when the system is slowed down, i.e., Windows Server, and it's awesome with 5 GB of memory usage. It's not a massive issue, and it's rare in testing after the updates just pushed. Like 1-40, I had it pop up, and that was as Finder files were moving and my CPU was pegged.

SenpaiHunters commented on Loop/Luminare/Settings/Keybindings/KeybindsConfigurationView.swift:
```diff
@@ -34,51 +18,77 @@ struct KeybindsConfigurationView: View {
+                    keybinds.insert(.init(.noAction), at: 0)
+                }
+
+                Button("Remove", role: .destructive) {
```
This does not accept the `del` key to remove keybinds; it also crashes when importing from the back file. See the attached image.

<img width="2562" height="356" alt="Image" src="https://github.com/user-attachments/assets/8018a5d1-4aec-410a-93a5-f410ceae7b6b" />

This is the `ContiguousArrayBuffer.swift: line 690 Fatal error: Index out of range`

SenpaiHunters replied:
Fixed the crash here; the delete key alone is not an issue. It’s just that cmd+delete seems like an extra step, but we can keep it as is. I just noticed that the `.keyboardShortcut(.delete)` was added, and that didn't do anything.

mrkai77 replied:
Think it's a good idea, since it will help prevent accidental deletions (#753).

SenpaiHunters commented on Loop/Luminare/Settings/Keybindings/KeybindsConfigurationView.swift:
```diff
@@ -34,51 +18,77 @@ struct KeybindsConfigurationView: View {
 
-            LuminareToggle("Double-click to trigger", isOn: $model.doubleClickToTrigger)
-            LuminareToggle("Middle-click to trigger", isOn: $model.middleClickTriggersLoop)
+            LuminareToggle("Double-click to trigger", isOn: $doubleClickToTrigger)
```
This doesn't seem to work in my testing; it could also be a personal issue. Please test to see locally if it does actually work.

mrkai77 replied:
Works on my side, mind explaining which part doesn't work?

SenpaiHunters replied:
This now works in testing, which is good. It seems there was some lag or issues on my end when testing this.

SenpaiHunters commented on Loop/Luminare/Settings/Keybindings/CustomActionConfigurationView.swift:
```diff
@@ -43,33 +45,42 @@ struct CustomActionConfigurationView: View {
     @ViewBuilder private func configurationSections() -> some View {
-        LuminareSection {
+        LuminareSection(outerPadding: 0) {
             LuminareTextField("Custom Keybind", text: Binding(get: { action.name ?? "" }, set: { action.name = $0 }))
```
If the popup asks if Loop can access the user's desktop file (we can actually remove this call), the display will not render unless the popup is accepted/confirmed. See both attached images.

If prompt is unaccepted:

<img width="1184" height="1042" alt="Image" src="https://github.com/user-attachments/assets/37677f8c-4332-4b69-9ecf-cad9837756d2" />

Once prompt is accepted and allowed access:

<img width="996" height="944" alt="Image" src="https://github.com/user-attachments/assets/d85c4136-7e86-4265-8da4-a80f8bf8810f" />

mrkai77 replied:
Great catch, made `WallpaperImageFetcher` to manage that

SenpaiHunters commented on Loop/Luminare/Loop/AdvancedConfiguration.swift:
```diff
@@ -129,25 +107,32 @@ struct AdvancedConfigurationView: View {
         LuminareSection("General") {
             if #available(macOS 15.0, *) {
-                LuminareToggle("Use macOS window manager when available", isOn: $model.useSystemWindowManagerWhenAvailable)
+                LuminareToggle("Use macOS window manager when available", isOn: $useSystemWindowManagerWhenAvailable)
```
When using the macOS window manager, the fullscreen snap, radial menu, or trigger menu will not work; it animates to indicate that it's going to move, yet fails to do so.
```
New preview window action received: maximize
Window snapping direction changed: maximize
Screen frame: (0.0, 0.0, 1440.0, 900.0)
Screen safeScreenFrame: (0.0, 25.0, 1440.0, 875.0)
Target frame: (0.0, 0.0, 1440.0, 875.0)
New preview window action received: maximize
Window snapping direction changed: maximize
Screen frame: (0.0, 0.0, 1440.0, 900.0)
Screen safeScreenFrame: (0.0, 25.0, 1440.0, 875.0)
Target frame: (0.0, 0.0, 1440.0, 875.0)
New preview window action received: maximize
Window snapping direction changed: maximize
Screen frame: (0.0, 0.0, 1440.0, 900.0)
Screen safeScreenFrame: (0.0, 25.0, 1440.0, 875.0)
Target frame: (0.0, 0.0, 1440.0, 875.0)
New preview window action received: maximize
Window snapping direction changed: maximize
Screen frame: (0.0, 0.0, 1440.0, 900.0)
Screen safeScreenFrame: (0.0, 25.0, 1440.0, 875.0)
Target frame: (0.0, 0.0, 1440.0, 875.0)
New preview window action received: maximize
Resizing Xcode to maximize on Built-in Retina Display
Target window frame: (0.0, 25.0, 1440.0, 875.0)
```

SenpaiHunters commented on Loop/Luminare/Loop/AdvancedConfiguration.swift:
```diff
@@ -129,25 +107,32 @@ struct AdvancedConfigurationView: View {
-                sliderRange: 5...50,
-                suffix: "px",
+            LuminareToggle(isOn: $animateWindowResizes) {
+                Text("Animate window resize")
```
Does this toggle still require the popover to indicate it's under development? We haven't touched it in a bit... and in most of my testing and day-to-day use, it's worked well. I have turned it off in the past 2 months, however, so I can't say anything here.

mrkai77 replied:
I think it's a good idea. While development on it has mostly stopped, it's also not quite as stable as I'd like, and I think it's important to mention that as well.

SenpaiHunters commented on Loop/Luminare/Settings/Behavior/BehaviorConfiguration.swift:
```diff
@@ -10,179 +10,139 @@ import Luminare
-                    info: SystemWindowManager.MoveAndResize.snappingEnabled ? model.systemSnappingWarning : nil,
-                    isOn: $model.windowSnapping
-                )
+                LuminareToggle(isOn: $windowSnapping) {
```
This rarely works now; it's about 1/5. It doesn't matter which direction of the screen; it takes a few tries to actually snap.

SenpaiHunters replied:
I found the issue. If the use of the macOS 15 default engine is turned on, it will not accept any keybinds it does not know. For this, we can add a fallback to revert to the loops engine when the default one is not registered.

SenpaiHunters reviewed (dismissed):
Hey, I reviewed the PR. There are a few things and crashes to fix; after that, we're good to merge!

SenpaiHunters committed:
🐞 Fixed some crashes

SenpaiHunters committed:
🔧 Update Luminare to remote - fix a crash - half fix text rendering

MrKai77 committed:
🔥 Remove Luminare submodule

MrKai77 committed:
🐞 Fix popover

MrKai77 committed:
🚛 Move ScreenView out of Luminare

MrKai77 committed:
✨ `WallpaperImageFetcher` to fetch wallpaper image

MrKai77 committed:
🐞 Pass event through if type is `tapDisabledByTimeout` or `tapDisabledByUserInput`

MrKai77 committed:
Window snapping on Loop, prevent larger/smaller actions on fixed size windows

MrKai77 committed:
We should take retained value from `CGSHWCaptureWindowList`

Otherwise, there is a memory leak. Also see: https://github.com/lwouis/alt-tab-macos/blob/b325cc75c02ea6685e9adef93e49a8a1700062fb/src/api-wrappers/CGWindowID.swift#L21

SenpaiHunters reviewed (dismissed):
I think one last thing we should do before squash merging this to main would be to fix the Swift format errors in the linting. After that, I think we're good to release.

MrKai77 committed:
🎨 Format
